### PR TITLE
[CMMD] Create a pipeline for CMMD dataset 

### PR DESCRIPTION
**Pull request by KacperRogala**
**Issue:** #80 

**Description:** 
Creates a preprocessing pipeline for images stored in CMMD dataset. 
The purpose is to add the ability to preprocess another medical image dataset to the repository. The dataset is CMMD (The Chinese Mammography Database) containing 5,202 MG (Mammography) breast images.

Additionally the file steps/convert_dcm2png.py is changed. Previously, for this step to work correctly, the masks_path argument had to be provided in the config/runner_config.py file, even if the dataset did not have masks. Now the masks_path argument is optional.

### CODE REVIEW

File: pyproject.toml
Comment (andrekomor) [inline pyproject.toml:1]: Is this package used anywhere? If no, please restore old poetry files.

File: src/pipelines/cmmd.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        # ensure uniqueness from the perspective of the entire dataset.
+        # Hence in the img_id construction it was decided to use the
+        # name of the folder one level higher. From the name in the
+        # format '1.000000-NA-xxxxx' (where xxxxx is a series of digits)
+        # xxxxx was used. Hence the unique img_id will have the structure
+        # xxxxx_1 or xxxxx_2.
+
+        lvl_higher = os.path.basename(os.path.dirname(img_path)).split("-")[2]
+
+        img_id_final = lvl_higher + "_" + img_id
```
Comment (andrekomor) [inline src/pipelines/cmmd.py:42]: Is `xxxxx` always same length? If so, then I think, we can omit the `+ "_" +` part. Adding another underscore breaks out of our naming convention, by adding another section to the file name, which might be problematic when users try to segregate the files by parsing their names in some way.

File: config/runner_config.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 """
 from src.base.pipeline import PathArgs
 from src.constants import TARGET_PATH
-from src.pipelines import (
+from src.pipelines import (  # CmmdPipeline,
```
Comment (andrekomor) [inline config/runner_config.py:11]: Fix imports. Move cmmd pipeline to import in brackets.

File: README.md [hunk 3]
```diff
@@ -243,6 +244,19 @@ Due to the copyright restrictions of the source datasets, we can't share the fil
...
 
 
+<details>
+  <summary>18. CMMD - The Chinese Mammography Database</summary>
+
+**18. CMMD**
+  1. Go to [CMMD](https://www.cancerimagingarchive.net/collection/cmmd/).
+  2. Download .tcia file from Data Access table (first row).
+  3. Download [NBIA Data Retriver](https://wiki.cancerimagingarchive.net/display/NBIA/Downloading+TCIA+Images) to be able to download images.
+  4. Download CMMD_clinicaldata_revision.xlsx from Data Access table (second row) for labels information.
```
Comment (andrekomor) [inline README.md:254]:
```suggestion
  2. Download .tcia file from Data Access table.
  3. Download [NBIA Data Retriver](https://wiki.cancerimagingarchive.net/display/NBIA/Downloading+TCIA+Images) to be able to download images.
  4. Download CMMD_clinicaldata_revision.xlsx from Data Access table for labels information.
```

Row positions, and data displaying might change.

File: README.md [hunk 4]
```diff
@@ -53,6 +53,7 @@ The labels and segmentation masks were unified to be compliant with RadLex ontol
 | 11 | [Chest X-ray 14](https://nihcc.app.box.com/v/ChestXray-NIHCC/folder/36938765345) | XRAY | classification|
 | 12 | [COCA- Coronary Calcium and chest CTs](https://stanfordaimi.azurewebsites.net/datasets/e8ca74dc-8dd4-4340-815a-60b41f6cb2aa) | CT | Segmentation |
 | 13 | [BrainMetShare](https://aimi.stanford.edu/brainmetshare) | MRI | Segmentation |
+| 18 | [CMMD](https://www.cancerimagingarchive.net/collection/cmmd/) | MG | Classification |
```
Comment (andrekomor) [inline README.md:56]:
```suggestion
| 18 | [CMMD](https://www.cancerimagingarchive.net/collection/cmmd/) | XRAY | Classification |
```

Mammography is x-ray I guess. Please confirm @A-Huli

File: config/dataset_config.py [hunk 5]
```diff
@@ -271,4 +271,17 @@ class DatasetArgs:
     },
     masks={masks.Metastasis.radlex_name: MaskColor(source_color=255, target_color=masks.Metastasis.color)},
 )
+
+cmmd = DatasetArgs(
+    dataset_uid="18",
+    dataset_name="cmmd",
+    phases={"0": "MG"},
```
Comment (andrekomor) [inline config/dataset_config.py:278]:
```suggestion
    phases={"0": "XRAY"},
```

Verify with @A-Huli

**Review by andrekomor — CHANGES_REQUESTED** [review]

File: pyproject.toml
Comment (KacperRogala) [reply pyproject.toml:1]: openpyxl package is required. Labels are saved in .xlsx file. To read them in LabelExtractor class pandas.read_excel() is used. Lack of openpyxl package causes error in this method.

File: README.md (hunk 4 above)
Comment (KacperRogala) [reply README.md:56]: After consultation with @A-Huli at today's meeting (02.09) it was decided to keep MG as the phase name, as the project does not yet have strict restrictions on the phase name ontology.

File: config/dataset_config.py (hunk 5 above)
Comment (KacperRogala) [reply config/dataset_config.py:278]: Similarly as above, after consultation with @A-Huli it was decided to keep MG as the phase name.

File: src/pipelines/cmmd.py (hunk 1 above)
Comment (KacperRogala) [reply src/pipelines/cmmd.py]: I changed it according to the suggestion.

File: config/runner_config.py (hunk 2 above)
Comment (KacperRogala) [reply config/runner_config.py]: Imports got fixed.

File: README.md (hunk 3 above)
Comment (KacperRogala) [reply README.md]: I changed it according to the suggestion.

File: pyproject.toml
Comment (andrekomor) [reply pyproject.toml:1]: Ok, closing

File: config/runner_config.py [hunk 6]
```diff
@@ -119,4 +120,11 @@
             target_path=TARGET_PATH,
         ),
     ),
+    CmmdPipeline(
+        path_args=PathArgs(
+            source_path="",  # Path to 'manifest-1616439774456/CMMD' folder
```
Comment (andrekomor) [inline config/runner_config.py:125]:
```suggestion
            source_path="",  # Path to 'manifest-{xxxxxxxxxxxx}/CMMD' folder
```
I think manifest number is unique to single process of downloading via tcia, not connected to the dataset.

Comment (andrekomor) [reply config/runner_config.py:125]: Also please fix pre-commit fails. There are some empty spaces left at the end of some lines. You can run it locally using `pre-commit run --all-files`

Comment (KacperRogala) [reply config/runner_config.py]: Changes are made.

Comment (KacperRogala) [reply config/runner_config.py]: I also had to change the `source_path` in` testing/tests/test_cmmd.py` a bit to match the path in the test_dummy_data folder on google drive.

File: testing/tests/test_cmmd.py [hunk 7]
```diff
@@ -15,7 +15,7 @@
 from src.pipelines.cmmd import CmmdPipeline
 from testing.libs.dataset_testing_lib import DatasetTestingLibrary
 
-source_path = os.path.join(os.getcwd(), "testing/test_dummy_data/18_cmmd/input/CMMD")
+source_path = os.path.join(os.getcwd(), "testing/test_dummy_data/18_cmmd/input/manifest-1616439774456/CMMD")
```
Comment (andrekomor) [inline testing/tests/test_cmmd.py:18]: I guess we could just leave it as it was. We are not really interested in manifest folder. we can have CMMD folder straight in input folder

**Review by andrekomor — APPROVED** [review]

File: testing/tests/test_cmmd.py (hunk 7 above)
Comment (KacperRogala) [reply testing/tests/test_cmmd.py]: I changed it back to "testing/test_dummy_data/18_cmmd/input/CMMD".

**Review by andrekomor — APPROVED** [review]

**Comment by andrekomor** [comment]
Make sure testing step is passing before merging

**Comment by KacperRogala** [comment]
> Make sure testing step is passing before merging

Honestly, I don't know why the test is failing. Based on the fail message, I understand that it can't find the `CMMD_clinicaldata_revision.xlsx` file. But it should be in the test files, at least it is on the drive.

**Pull request merged**